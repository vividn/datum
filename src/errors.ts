export class MyError extends Error {
  constructor(m?: unknown) {
    if (typeof m === "string") {
      super(m);
    } else {
      super();
    }
    this.name = this.constructor.name;
    Object.setPrototypeOf(this, MyError.prototype);
  }
}

export class MissingRequiredKeyError extends MyError {
  constructor(key: string) {
    super(`No data given for the required key: ${key}`);
    Object.setPrototypeOf(this, MissingRequiredKeyError.prototype);
  }
}

export class ExtraDataError extends MyError {
  constructor(keys: string[]) {
    super(
      `the following data do not have keys: ${keys.join(
        ", ",
      )}. Assign keys with equals signs, use required/optional keys, specify a key to use as --remainder, or use --lenient`,
    );
    Object.setPrototypeOf(this, ExtraDataError.prototype);
  }
}
export class BaseDataError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, BaseDataError.prototype);
  }
}

export class IdError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, IdError.prototype);
  }
}

export class BadTimeError extends MyError {
  #key = "";
  #value = "";

  constructor(badTime?: string) {
    super();
    this.#value = badTime ?? "";
    Object.setPrototypeOf(this, BadTimeError.prototype);
  }

  get message(): string {
    const withValue = this.#value ? `"${this.#value}" is an ` : "";
    const withKey = this.#key ? `, key: ${this.#key}` : "";
    return `${withValue}invalid time${withKey}`;
  }
  set key(key: string) {
    this.#key = key;
  }
  get key(): string {
    return this.#key;
  }
  get value(): string {
    return this.#value;
  }
}

export class BadDateError extends MyError {
  #key = "";
  #value = "";

  constructor(badDate?: string) {
    super();
    this.#value = badDate ?? "";
    Object.setPrototypeOf(this, BadDateError.prototype);
  }

  get message(): string {
    const withValue = this.#value ? `"${this.#value}" is an ` : "";
    const withKey = this.#key ? `, key: ${this.#key}` : "";
    return `${withValue}invalid date${withKey}`;
  }
  set key(key: string) {
    this.#key = key;
  }
  get key(): string {
    return this.#key;
  }
  get value(): string {
    return this.#value;
  }
}

export class BadDurationError extends MyError {
  #key = "";
  #value = "";

  constructor(badDuration?: string) {
    super();
    this.#value = badDuration ?? "";
    Object.setPrototypeOf(this, BadDurationError.prototype);
  }

  get message(): string {
    const withValue = this.#value ? `"${this.#value}" is an ` : "";
    const withKey = this.#key ? `, key: ${this.#key}` : "";
    return `${withValue}invalid duration${withKey}. Give a valid duration string or skip duration with a "."`;
  }
  set key(key: string) {
    this.#key = key;
  }
  get key(): string {
    return this.#key;
  }
  get value(): string {
    return this.#value;
  }
}

export class BadTimezoneError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, BadTimezoneError.prototype);
  }
}

export class BadStateError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, BadStateError.prototype);
  }
}

export class MigrationError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, MigrationError.prototype);
  }
}

export class MergeError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, MergeError.prototype);
  }
}

export class NotImplementedError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, NotImplementedError.prototype);
  }
}

export type CouchDbError = {
  error: string;
  reason: string;
  status: number;
  name: string;
  message: string;
};

export const isCouchDbError = (error: unknown): error is CouchDbError => {
  return !!(error as CouchDbError).error && !!(error as CouchDbError).reason;
};

export class DatumViewMissingError extends MyError {
  constructor(map_name?: unknown) {
    super(
      `Missing internal datum view ${map_name}. Please run setup on this database`,
    );
    Object.setPrototypeOf(this, DatumViewMissingError.prototype);
  }
}

export class LastDocsTooOldError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, LastDocsTooOldError.prototype);
  }
}

export class FieldError extends MyError {
  constructor(m: unknown) {
    super(m);
    Object.setPrototypeOf(this, FieldError.prototype);
  }
}
