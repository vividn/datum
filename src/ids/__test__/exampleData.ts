import { DatumData, DatumMetadata } from "../../documentControl/DatumDocument";

export const exampleData: DatumData = {
  foo: "abc",
  bar: "def",
  complex: { data: "nested" },
  array: ["various", 2, "data"],
  num: 3,
  "wei%rd": "da%ta",
};
export const exampleOccurTime = "2020-11-09T00:35:10.000Z";
export const exampleDataOccur = {
  occurTime: exampleOccurTime,
  occurUtcOffset: 1,
  ...exampleData,
};
export const exampleDataOccurField: DatumData = {
  ...exampleDataOccur,
  field: "main",
};
export const exampleMeta: DatumMetadata = {
  createTime: "2020-11-09T00:40:12.544Z",
  modifyTime: "2020-11-09T00:40:12.544Z",
  random: 0.7368733800261729,
  humanId: "mqp4znq4cvp3qnj74fgi9",
};
