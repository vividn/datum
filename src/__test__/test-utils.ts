import { CouchDbError } from "../errors";
import {
  DataOnlyDocument,
  DatumData,
  DatumDocument,
  DatumMetadata,
  EitherDocument,
  EitherPayload,
} from "../documentControl/DatumDocument";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "../time/parseTimeStr";
import { now } from "../time/timeUtils";
import { connectDb } from "../auth/connectDb";
import * as connectDbModule from "../auth/connectDb";
import { defaultIdComponents } from "../ids/defaultIdComponents";
import { buildIdStructure } from "../ids/buildIdStructure";
import { defaults } from "../input/defaults";
import { assembleId } from "../ids/assembleId";
import * as newHumanIdModule from "../meta/newHumanId";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const mockDocMissingError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing",
  message: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  status: 404,
  message: "deleted",
  name: "not_found",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  status: 404,
  name: "not_found",
  error: "not_found",
  reason: "missing_named_view",
  message: "missing_named_view",
};

export async function resetTestDb(
  db: PouchDB.Database & {
    __opts?: PouchDB.Configuration.DatabaseConfiguration;
    _destroyed?: boolean;
    _closed?: boolean;
  },
): Promise<PouchDB.Database> {
  await db.destroy().catch(pass);
  // nasty hack to reopen closed database
  delete db._destroyed;
  delete db._closed;
  return db.constructor(db.name, db.__opts);
}

export function testDbLifecycle(
  dbName: string,
): PouchDB.Database<EitherPayload> {
  const db = connectDb({ db: dbName });

  beforeEach(async () => {
    await resetTestDb(db);
    jest.spyOn(connectDbModule, "connectDb").mockReturnValue(db);
  });

  afterEach(async () => {
    await db.destroy().catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr, referenceTime: DateTime.now() });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => Promise<O>,
): (...args: A) => Promise<O> {
  return async (...args: A): Promise<O> => {
    pushNow(timeStr);
    const returnVal = await fn(...args);
    popNow();
    return returnVal;
  };
}

// TODO: Transition all tests to use makeDoc where appropriate
export function makeDoc(
  data: DatumData,
  meta: DatumMetadata | false = {},
  include_rev = false,
): EitherDocument {
  let doc: EitherDocument;
  if (meta === false) {
    doc = { ...data } as DataOnlyDocument;
  } else {
    doc = { ...data, meta } as DatumDocument;
  }

  meta = meta || {};
  if (doc._id === undefined) {
    let idStructure: string;
    if (meta?.idStructure) {
      idStructure = meta.idStructure;
    } else {
      const { defaultIdParts, defaultPartitionParts } = defaultIdComponents({
        data,
      });
      idStructure = buildIdStructure({
        idParts: defaultIdParts,
        delimiter: defaults.idDelimiter,
        partition: defaultPartitionParts,
      });
    }
    doc._id = assembleId({ payload: doc, idStructure });
  }

  if (include_rev) {
    doc._rev = "1-foo";
  }
  return doc;
}

export async function deterministicHumanIds(seed?: number): Promise<void> {
  let a = seed || 20231018;
  function random() {
    const x = Math.sin(a++) * 10000;
    return x - Math.floor(x);
  }
  function mockNewHumanId(): string {
    return random().toString(36).slice(2) + random().toString(36).slice(2);
  }

  beforeEach(() => {
    a = seed || 20231018;
    jest
      .spyOn(newHumanIdModule, "newHumanId")
      .mockImplementation(mockNewHumanId);
  });
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
