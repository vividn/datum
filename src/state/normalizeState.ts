import { BadStateError } from "../errors";
import { JsonObject, JsonType } from "../utils/utilityTypes";
import isPlainObject from "lodash.isplainobject";
import flattenDeep from "lodash.flattendeep";

export type StateObject = JsonObject & {
  id: string | boolean;
};
export type SingleState = string | boolean | StateObject;
export type DatumState = null | SingleState | SingleState[];

export function isStateObject(state: SingleState): state is StateObject {
  return isPlainObject(state);
}
export function normalizeState(state: JsonType): DatumState {
  if (state === null) {
    return null;
  }
  if (typeof state === "number") {
    return String(state);
  }
  if (state === "start") {
    return true;
  }
  if (state === "end") {
    return false;
  }
  if (typeof state === "string" || typeof state === "boolean") {
    return state;
  }
  if (Array.isArray(state)) {
    if (state.length === 0) {
      return false;
    }
    if (state.length === 1) {
      return normalizeState(state[0]);
    }
    return flattenDeep(
      state.map((innerState) => {
        const normalized = normalizeState(innerState);
        if (normalized === null) {
          throw new BadStateError(
            "null is a special state to indicate that the field is not being tracked and cannot exist together with other states",
          );
        }
        return normalized;
      }),
    );
  }
  if (state.id === undefined) {
    state.id = true;
  }
  const id = state.id as JsonType;
  const otherKeys = { ...state };
  delete otherKeys.id;

  if (Object.keys(otherKeys).length === 0) {
    return normalizeState(id);
  }
  const normalizedId = normalizeState(id);
  if (normalizedId === null) {
    if (Object.keys(otherKeys).length > 0) {
      console.warn(
        "null is a special state to indicate that the field is not being tracked, but additional state data was provided. It has been removed",
      );
    }
    return null;
  }
  if (typeof normalizedId === "string" || typeof normalizedId === "boolean") {
    return { id: normalizedId, ...otherKeys };
  }

  if (Array.isArray(normalizedId)) {
    return normalizedId.map((single) => {
      if (typeof single === "string" || typeof single === "boolean") {
        return { id: single, ...otherKeys };
      }
      return { ...single, ...otherKeys };
    });
  }
  return { ...normalizedId, ...otherKeys };
}
