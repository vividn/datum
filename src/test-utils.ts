import Nano, { DocumentScope } from "nano";
import { CouchDbError } from "./errors";
import { EitherPayload } from "./documentControl/DatumDocument";
import * as connectDb from "./auth/connectDb";
import Mock = jest.Mock;
import { DateTime, Settings } from "luxon";
import { parseTimeStr } from "./time/parseTimeStr";
import { now } from "./time/timeUtils";

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const pass = (): void => {};
export const fail = (): never => {
  throw Error;
};

export const testNano = Nano(`http://admin:password@localhost:5983`);

export const mockDocMissingError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing",
  error: "not_found",
  reason: "missing",
};

export const mockDocDeletedError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "deleted",
  error: "not_found",
  reason: "deleted",
};

export const mockMissingNamedViewError: CouchDbError = {
  scope: "couch",
  statusCode: 404,
  errid: "non_200",
  description: "missing_named_view",
  error: "not_found",
  reason: "missing_named_view",
};

export async function resetTestDb(
  dbName: string
): Promise<DocumentScope<EitherPayload>> {
  const maxTries = 3;
  let tries = 0;
  await testNano.db.destroy(dbName).catch(pass);
  while (tries++ <= maxTries) {
    await testNano.db.destroy(dbName).catch(pass);
    await testNano.db.create(dbName).catch(pass);
    if ((await testNano.db.list()).includes(dbName)) {
      return testNano.use(dbName) as DocumentScope<EitherPayload>;
    }
  }
  throw Error(`Unable to reset database after ${maxTries} attempts`);
}

export function testDbLifecycle(dbName: string): DocumentScope<EitherPayload> {
  const db = testNano.use(dbName) as DocumentScope<EitherPayload>;

  beforeEach(async () => {
    await resetTestDb(dbName);
    jest.spyOn(connectDb, "connectDb").mockImplementation(async () => db);
  });

  afterEach(async () => {
    await testNano.db.destroy(dbName).catch(pass);
  });

  return db;
}

export function mockedLogLifecycle(): Mock {
  const originalLog = console.log;
  const mockedLog = jest.fn() as Mock;

  beforeEach(async () => {
    console.log = mockedLog;
  });

  afterEach(async () => {
    console.log = originalLog;
    mockedLog.mockReset();
  });

  return mockedLog;
}

const originalNowFn = Settings.now;
const nowStack: (() => number)[] = [];
export function setNow(timeStr: string): DateTime {
  const parsedTime = parseTimeStr({ timeStr });
  const newNow = () => parsedTime.toMillis();
  Settings.now = newNow;
  return parsedTime;
}

export function pushNow(timeStr: string): DateTime {
  nowStack.push(Settings.now);
  return setNow(timeStr);
}

export function popNow(): DateTime {
  const lastNow = nowStack.pop();
  if (lastNow === undefined) {
    throw new Error("tried to pop non existent now");
  }
  Settings.now = lastNow;
  return now();
}

export function restoreNow(): DateTime {
  Settings.now = originalNowFn;
  nowStack.length = 0;
  return now();
}

export function at<A extends any[], O>(
  timeStr: string,
  fn: (...args: A) => O
): (...args: A) => O {
  return (...args: A): O => {
    pushNow(timeStr);
    const returnVal: O = fn(...args);
    popNow();
    return returnVal;
  };
}

// export async function generateSampleDay(dateStr = "2022-08-14") {
//
//   setNow(dateStr);
//   at('8:30', addCmd)({field: "sleep", })
//
//   popNow();
// }
