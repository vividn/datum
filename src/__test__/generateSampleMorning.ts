import { endCmd } from "../commands/endCmd";
import { occurCmd } from "../commands/occurCmd";
import { startCmd } from "../commands/startCmd";
import { switchCmd } from "../commands/switchCmd";
import { popNow, pushNow, setNow } from "./test-utils";

export async function generateSampleMorning(date: string): Promise<void> {
  pushNow(`8:30 ${date}`);
  await endCmd("sleep");
  setNow("+5");
  await startCmd("sleep -y -t 2320");
  setNow("8:40");
  await switchCmd("project german");
  await switchCmd("text fiction_book");
  setNow("9:10");
  await endCmd("project");
  await endCmd("text");
  setNow("9:30");
  await switchCmd("environment outside");
  await startCmd("stretch");
  setNow("+7");
  await endCmd("stretch");
  await startCmd("run");
  setNow("+30");
  await endCmd("run distance=5.4");
  await startCmd("stretch");
  setNow("+8");
  await endCmd("stretch");
  await switchCmd("environment home");
  setNow("+3");
  await occurCmd("pushup amount=10");
  setNow("10:30");
  await switchCmd("environment outside 5");
  setNow("11");
  await occurCmd("caffeine amount=100 -c coffee");
  popNow();
}
