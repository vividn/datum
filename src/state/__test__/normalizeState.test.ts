import { normalizeState } from "../normalizeState";
import { mockedLogLifecycle } from "../../__test__/test-utils";
import { BadStateError } from "../../errors";

describe("unfoldState", () => {
  const { mockedWarn } = mockedLogLifecycle();

  it("returns null if state is null", () => {
    expect(normalizeState({ id: null })).toEqual(null);
    expect(normalizeState(null)).toEqual(null);
  });

  it("returns non object state without modification", () => {
    expect(normalizeState("abcd")).toEqual("abcd");
    expect(normalizeState(false)).toEqual(false);
    expect(normalizeState(true)).toEqual(true);
  });

  it("stringifys number states", () => {
    expect(normalizeState(123)).toEqual("123");
  });

  it("unfolds objects with just id into their simple value counterparts", () => {
    expect(normalizeState({ id: true })).toEqual(true);
    expect(normalizeState({ id: false })).toEqual(false);
    expect(normalizeState({ id: 123 })).toEqual("123");
    expect(normalizeState({ id: null })).toEqual(null);
  });

  it("keeps an object state with an id as it is", () => {
    expect(normalizeState({ id: "id", abc: 123, key: "value" })).toEqual({
      id: "id",
      abc: 123,
      key: "value",
    });
  });

  it("assigns a state id of true if it is not defined in an object", () => {
    expect(normalizeState({ abc: 123, key: "value" })).toEqual({
      id: true,
      abc: 123,
      key: "value",
    });
    expect(normalizeState({})).toEqual(true);
  });

  it("unfolds an id into the state if id is an object", () => {
    expect(normalizeState({ id: { abc: 123, key: "value" } })).toEqual({
      id: true,
      abc: 123,
      key: "value",
    });
    expect(normalizeState({ id: { id: "abc" } })).toEqual("abc");
    expect(
      normalizeState({ id: { id: "stringId", abc: 123, key: "value" } }),
    ).toEqual({
      id: "stringId",
      abc: 123,
      key: "value",
    });
  });

  it("interpolates other fields into an unfolded id, overwriting as necessary", () => {
    expect(
      normalizeState({
        id: { id: "stringId", abc: 123, key: "value" },
        key: "replacementValue",
      }),
    ).toEqual({
      id: "stringId",
      abc: 123,
      key: "replacementValue",
    });
  });

  it("recursively unfolds nested ids", () => {
    expect(
      normalizeState({
        id: {
          id: { id: "deepId", key: "value1", abc: 123, untouched: "inner_key" },
          key: "value2",
          nested: { inner: "key" },
        },
        key: "value3",
        abc: 456,
      }),
    ).toEqual({
      id: "deepId",
      key: "value3",
      abc: 456,
      untouched: "inner_key",
      nested: { inner: "key" },
    });
  });

  it("recursively normalizes an array state", () => {
    expect(
      normalizeState(["abcd", { id: "stringId" }, true, { abc: "def" }]),
    ).toEqual(["abcd", "stringId", true, { id: true, abc: "def" }]);
  });

  it("keeps a complex state with id=true the same", () => {
    expect(normalizeState({ id: true, name: "name", n: 3 })).toEqual({
      id: true,
      name: "name",
      n: 3,
    });
  });

  it("turns a complex state with id=null into just {null, and warns", () => {
    expect(normalizeState({ id: null, name: "name", n: 3 })).toEqual(null);
    expect(mockedWarn).toHaveBeenCalledTimes(1);
    expect(mockedWarn.mock.calls[0][0]).toMatchSnapshot();
  });

  it("turns an array of length 0 into a false state", () => {
    expect(normalizeState([])).toEqual(false);
    expect(normalizeState({ id: [] })).toEqual(false);
    expect(normalizeState({ id: [], name: "john" })).toEqual({
      id: false,
      name: "john",
    });
  });

  it("unfolds an object with an array of ids into an array of objects with the other fields in all of them", () => {
    expect(normalizeState({ id: ["state1", "state2", "state3"] })).toEqual([
      "state1",
      "state2",
      "state3",
    ]);
    expect(
      normalizeState({
        id: ["state1", "state2", true],
        name: "name",
        n: 3,
      }),
    ).toEqual([
      { id: "state1", name: "name", n: 3 },
      { id: "state2", name: "name", n: 3 },
      { id: true, name: "name", n: 3 },
    ]);
  });

  it("flattens nested arrays of state", () => {
    expect(
      normalizeState([
        { id: ["state1", "state2", "state3"] },
        { id: "state4" },
        "state5",
      ]),
    ).toEqual(["state1", "state2", "state3", "state4", "state5"]);
    expect(
      normalizeState({
        id: [
          "state1",
          { id: "state2" },
          { id: ["state3", ["state4", "state5"]], name: "john" },
        ],
      }),
    ).toEqual([
      "state1",
      "state2",
      { id: "state3", name: "john" },
      { id: "state4", name: "john" },
      { id: "state5", name: "john" },
    ]);
  });

  it("turns an input array of length 1 into an non array", () => {
    expect(normalizeState(["abcd"])).toEqual("abcd");
    expect(normalizeState([{ id: "abcd", name: "john" }])).toEqual({
      id: "abcd",
      name: "john",
    });
    expect(normalizeState([{ id: null }])).toEqual(null);
    expect(normalizeState([null])).toEqual(null);
  });

  it("throws an error if a null state is part of an array of states", () => {
    expect(() => normalizeState(["abc", null])).toThrow(BadStateError);
    expect(() => normalizeState([null, "abc"])).toThrow(BadStateError);
    expect(() => normalizeState([null, null])).toThrow(BadStateError);
    expect(() => normalizeState(["abc", { id: null }])).toThrow(BadStateError);
    expect(() => normalizeState([null, { id: "abc" }])).toThrow(BadStateError);
    expect(() => normalizeState([{ id: "abc" }, { id: null }])).toThrow(
      BadStateError,
    );
  });

  it("converts a state of 'start' to true", () => {
    expect(normalizeState("start")).toBe(true);
    expect(normalizeState(["a", "start"])).toEqual(["a", true]);
    expect(normalizeState({ id: "start", extra: "key" })).toEqual({
      id: true,
      extra: "key",
    });
  });

  it("converts a state of 'end' to false", () => {
    expect(normalizeState("end")).toBe(false);
    expect(normalizeState(["a", "end"])).toEqual(["a", false]);
    expect(normalizeState({ id: "end", extra: "key" })).toEqual({
      id: false,
      extra: "key",
    });
  });
});
