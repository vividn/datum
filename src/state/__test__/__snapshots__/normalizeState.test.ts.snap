// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unfoldState turns a complex state with id=null into just {null, and warns 1`] = `"null is a special state to indicate that the field is not being tracked, but additional state data was provided. It has been removed"`;
